### ISSUE TITLE
IP Blocked from Newegg

### ISSUE THREAD

**Issue by headbasha** [Ataraksia/NeweggBot#26]
Do you have time to update with a random interval to add items to the cart for purchase? It took only about an hour to get IP blocked from newegg for 24 hours.

**Comment by retrac752** [comment]
I also got ipblocked by newegg for 24 hours after running the bot for about 2 hours

**Comment by retrac752** [comment]
i mean to make the interval random, I'm replacing	
await page.waitForTimeout(config.refresh_time * 1000)
with	
await page.waitForTimeout((config.refresh_time + (Math.random() * config.refresh_time * 2)) * 1000)
this also makes the refresh up to 3x slower to make it seem more human like, but would this really be enough to trick newegg?

**Comment by headbasha** [comment]
I usually get IP banned within an hour but im also looking at about 15 different newegg products trying to buy a 3080. Im having to use my VPN to just keep refreshing IP's. Ill try that random number string and see how soon it IP bans.

**Comment by mateuszdrab** [comment]
I had to put that line somewhere else but it doesnt work for me totally.
Est total field shows empty and the script doesn't know what to do.

**Comment by Codolophier** [comment]
> this also makes the refresh up to 3x slower to make it seem more human like, but would this really be enough to trick newegg?

@retrac752  This is how I've been running. I placed a comment in Issue #12 about 10 days ago with the same pattern to it.  I've been running my bot almost 24x7 since then , no ban so far :) I pushed the code, along with some other changes, to my fork earlier today.  I'm not sure if @Ataraksia is watching this repo anymore, but I submitted a PR about 2 weeks ago that contained updates to accommodate for changes in the checkout flow by Newegg.

**Comment by retrac752** [comment]
@Codolophier I changed my code to randomize the time but also noticed and fixed the same bug you did where the bot would check the url for "ShoppingCart" instead of "cart" which would always fail, so we would never actually enter the check cart function to call the delay, so I'm unsure if it was the fact that this bot was refreshing without delay or if the randomization actually helped

I've also ran the bot for almost 24 hours with no problems since this fix. However, since I've never entered the checkout flow, I didn't realize there was an issue there. Have you successfully checked out of newegg with your code changes that looks for the secure checkout button?

**Comment by mateuszdrab** [comment]
I pretty much had to rewrite the whole thing. Nothing worked, it seems the UI changed or something. I had to make it go through each step of the checkout and then hit place order but at this point I think my credit card will kill it all with some verifications

**Comment by Codolophier** [comment]
@retrac752 
> Have you successfully checked out of newegg with your code changes that looks for the secure checkout button?
Yes, I've specified an item number that is in-stock and the bot successfully completed the purchase.  

> Also, is chrome better than firefox for this?
Either should be fine.

**Comment by Codolophier** [comment]
@mateuszdrab Are your account defaults all set for payment method as well as the billing and shipping address?  When I first tried the bot, I was getting all kinds of prompts to select and confirm things.  Once I set all of my defaults, the only input required was the CVV.

**Comment by mateuszdrab** [comment]
@Codolophier yes the defaults are set but since I've never purchased anything from them that might be why.
Can you cancel an order straight after placing it?

**Comment by Codolophier** [comment]
> Can you cancel an order straight after placing it?

@mateuszdrab Yes, you can. If you want to test the bot out, change the **auto_submit** value in the **config.json** to **false**.  This way, the bot will got all the way up to the CVV input, but won't proceed any further.  

**A note on the Newegg account defaults**:  Originally, I had some problems with the flow.  I thought that all of my defaults were set, but I was still getting prompted for payment and delivery method.   Long story short, I had to go through my account settings to the Address Book and give a name to my default address.  Then, in the Payment Options, despite having my credit card as the default payment, I removed any other payments types - such as PayPal.  Also under the Payment Options, under the selected card, I selected the **Select from Address **Book** option in order to set my Billing Address to the one I had saved earlier and given a name to.  I then gave the card a name, clicked the **Make Default** box, then the checkout flow no longer prompted me for additional information.

**Comment by technaustin** [comment]
> N82E16820313309

I just did this test with a cheap item, and set to false, and it did work once, for a cheaper item. When I tried it with something more expensive, I ran into the same problem I have been having with GFX cards. Here is a snapshot of the N95 pop up, with my PowerShell Window. It looks like sometimes, this window pops up and interrupts the bot.

![image](https://user-images.githubusercontent.com/46908628/102127825-d5ee8480-3e01-11eb-84e4-28bd3b02e81d.png)

**Comment by Codolophier** [comment]
@technaustin If I can reproduce - either directly or via unit test -  I can add some code to detect the popup and click the **I'm not interested** button.  I'm currently working on a new version that should be capable of handling these type of interruptions.  It may take me a few days (day job gets in the way sometimes :)), but I'll do what I can as soon as possible. 

PS: If you see this modal popup again, could you send the HTML for the popup/page?  It would help to use that as a model for the compensation code if I cannot reproduce those results.

**Comment by technaustin** [comment]
> @technaustin If I can reproduce - either directly or via unit test - I can add some code to detect the popup and click the **I'm not interested** button. I'm currently working on a new version that should be capable of handling these type of interruptions. It may take me a few days (day job gets in the way sometimes :)), but I'll do what I can as soon as possible.
> 
> **PS**: If you - or anyone else - encounters this modal popup again, could you send the HTML for the popup/page? It would help to use that as a model for the compensation code if I cannot reproduce those results.

Awesome! I’ll get it to you tomorrow morning. I can force it to happen again.

**Comment by mateuszdrab** [comment]
I fixed most of the issues in this script, took me like 4 hours to get it to checkout to the last stage since for me Newegg doesn't auto skip to review page on purchase. Still, I decided not to buy from Newegg since they're trash to me due to the combos.

```
     try {
                await page.waitForSelector('button#Masks_addtocart.btn.btn-primary' , {timeout: 10000})
                await page.click('button.btn[data-dismiss="modal"]')
                await report('Skipped stupid mask advert')
                }catch(err) {
                await report('when skipping mask')
                await report(err)
        }
```

**Comment by dexmar** [comment]
@mateuszdrab did you fork the repository or just working locally?

**Comment by mateuszdrab** [comment]
@dexmar, nope, but I can do so and push my changes back to it

**Comment by dexmar** [comment]
@mateuszdrab that would be great! both interested in landing a 3080 (still!) and interested in learning Node. thank you!!

**Comment by mateuszdrab** [comment]
@dexmar https://github.com/mateuszdrab/NeweggBot

It's messy, I am in no way a Node dev ;)

**Comment by dexmar** [comment]
wow, lightning fast! thank you! @mateuszdrab

**Comment by mazukou87** [comment]
@Codolophier Using yours with an item thats in stock (nvme ssd) to test it, it puts it in the cart, and then just keeps adding more and never actually goes to the checkout

Also did what you said about the defaults and its still throwing the error in the script and pausing everything, not sure what else to try

**Comment by Codolophier** [comment]
@mazukou87 It looks like the problem is that Newegg changed - or at least is varying - the resulting URL for the cart when an item has been added.  The code is currently checking for **cart**.  From some tests I've done, I can see that sometimes it is **cart**, other times **Cart**.  To fix this, change the line that reads:

```typescript
if (page.url().includes("cart")) {
 ...
}
```
to

```typescript
if (page.url().toLowerCase().includes("cart")) {
...
}
```

**Comment by mazukou87** [comment]
> ```ts
> if (page.url().toLowerCase().includes("cart")) {
> ```

did that, still keeps adding,
. I never really learned much html or java script, spent most of my time on c# and ruby, but that part looks like it would be right, but unless im missing something, which clearly i am, it seems like its noticing its on the cart but keeps going. Would a better option be to check the "x Item" line and look for a non zero integer? That at least least seems to be consistent

looking thru the else if statement, that seems to be the only thing i notice, if cart is found in url, then go to checkout,, but it seems to break at the cart check

**Comment by Codolophier** [comment]
@mazukou87 
> hope it doesnt sound like im being unappreciative or nit picking, this is just how i talk, very blunt and to the point haha

Don't sweat it, we're all here working together, honesty/directness is always best in programming terms :).  I tested the changes using an NVME with Item # N82E16820156245.  Things are working on this end.  You may want to try debugging the code, stepping through from the point the item is added to the cart to see where it is dumping out - or go old school and put a bunch of **console.log** statements everywhere (step debugging is the easiest though... and less cleanup when done) :)

**Comment by mazukou87** [comment]
> 
> 
> @mazukou87
> 
> > hope it doesnt sound like im being unappreciative or nit picking, this is just how i talk, very blunt and to the point haha
> 
> Don't sweat it, we're all here working together, honesty/directness is always best in programming terms :). I tested the changes using an NVME with Item # N82E16820156245. Things are working on this end. You may want to try debugging the code, stepping through from the point the item is added to the cart to see where it is dumping out - or go old school and put a bunch of **console.log** statements everywhere (step debugging is the easiest though... and less cleanup when done) :)

how do i step into javascript. i tried doing like i do c# in VS, but it just tells me to install some firefox package...despite it already being installed. This is my first foray in to JS lol

**Comment by Codolophier** [comment]
It depends on your IDE, but typically it isn't so bad.  I usually use Visual Studio Code (VSCode - free and "easy").  Here are some articles to follow for both VS Code and Visual Studio proper.

- https://code.visualstudio.com/docs/nodejs/nodejs-debugging
- https://docs.microsoft.com/en-us/visualstudio/javascript/debug-nodejs?view=vs-2019
